### ISSUE TITLE
[szip:x64-mingw-dynamic] Reason for `Could not find proper second linker member`

### ISSUE THREAD

**Issue by thumpro** [microsoft/vcpkg#14322]
I am playing with my version of the vcpkg, as per this https://github.com/microsoft/vcpkg/issues/10349 issue, to make it bootstrap under MSYS2-MINGW64 without using any Visual Studio related toolchain. I changed the relevant parts of the bootstrap scripts and made the C++ codes compile with g++ etc. Now I have a vcpkg version bootstrapped under MSYS2 running fine. Most of the ports I am interested in are also building just fine, keep in mind that I am using community triplet `x64-mingw-dynamic` and it is not officially supported in the first place. 

But while building `szip` as part of `opencv`:

> ./vcpkg.exe install szip:x64-mingw-dynamic

>Your feedback is important to improve Vcpkg! Please take 3 minutes to complete our survey by running: vcpkg contact --survey
Computing installation plan...
The following packages will be built and installed:
    szip[core]:x64-mingw-dynamic
Detecting compiler hash for triplet x64-mingw-dynamic...
Starting package 1/1: szip:x64-mingw-dynamic
Building package szip[core]:x64-mingw-dynamic...
Could not locate cached archive: C:\Users\burak\AppData\Local\vcpkg\archives\78\789f979763e4fb257ce37380862d2c721a65fb74.zip
-- Using community triplet x64-mingw-dynamic. This triplet configuration is not guaranteed to succeed.
-- [COMMUNITY] Loading triplet configuration from: C:\dev\bin\msys64\opt\vcpkg\triplets\community\x64-mingw-dynamic.cmake
-- Using cached C:/dev/bin/msys64/opt/vcpkg/downloads/szip-2.1.1.tar.gz
-- Cleaning sources at C:/dev/bin/msys64/opt/vcpkg/buildtrees/szip/src/szip-2.1.1-6b767acfa7.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/dev/bin/msys64/opt/vcpkg/downloads/szip-2.1.1.tar.gz
-- Applying patch fix-szip-config-to-set-szip-found.patch
-- Using source at C:/dev/bin/msys64/opt/vcpkg/buildtrees/szip/src/szip-2.1.1-6b767acfa7.clean
-- Configuring x64-mingw-dynamic
-- Building x64-mingw-dynamic-dbg
-- Building x64-mingw-dynamic-rel
-- Performing post-build validation
Could not find proper second linker member

I want to know what could be the reason for `Could not find proper second linker member` message since actual build proceeds without errors. I didn't modify `cofffilereader.cpp` to make it bootstrap under MSYS2 and it was validating other ports just fine. If someone could point me towards the correct direction as to cause of that error msg, I would be glad. 

I found a ticket on a similar issue: 
https://github.com/microsoft/vcpkg/pull/10398

but I don't know how to make use of it.

I also used `dumpbin` myself on the generated lib file. It indeed shows:
>36C size
correct header end
    34 public symbols

for `dumpbin /linkermember:1 szip_D.lib` but not for `dumpbin /linkermember:2 szip_D.lib`. The other lib files I generated with my `vcpkg` don't have such problems. Is it maybe `szip` is not viable to building as a dynamic library?

**Comment by sfhacker** [comment]
@thumpro Have you contacted us recently about you having issues with other ports?
You should refrain yourself from mixing environments and/or compilers! You are looking for trouble.
Just built szip (dynamic) from source on Windows 10 without issues.

![image](https://user-images.githubusercontent.com/25817529/97856742-d26ed480-1d50-11eb-99c5-d28a0e5e236e.png)

> 02/11/2020  09:16 PM            87,999 szip.dll

It is a bit hard to follow your description of this issue! Can you further elaborate on it?

**Comment by thumpro** [comment]
Note that this issue has nothing to do with the previous one I opened about a build failure on `x64-mingw-dynamic`. [https://github.com/microsoft/vcpkg/issues/14012]

This time, I am trying the investigate a feature request found on following open issue: https://github.com/microsoft/vcpkg/issues/10349. There they suggest making vcpkg bootstrappable using MinGW64 toolchain and outline the stage where they failed to compile vcpkg itself,

To that end, I modified the vcpkg source code and finally bootstrapped it under MinGW64. After, I wanted to check by building some ports to see if it is correctly working. I was able to build many but `szip` failed during post-build check with the error `Could not find proper second linker member`. I am not keen on modifying the source file `cofffilereader.cpp` myself, so I was asking for explanation on what this error message could mean.

**Comment by sfhacker** [comment]
The issue #10349 is from March and it remains open! Wow!
I didn't have/need to modify any file in order to build szip successfully.
Probably your environment is broken somehow.

**Comment by Neumann-A** [comment]
`set(VCPKG_POLICY_SKIP_DUMPBIN_CHECKS enabled)` skips the dumpbin checks.  second linker member are only generated if VS link.exe is used

**Comment by thumpro** [comment]
@Neumann-A That's strange. I will try it as soon as I have the time. It is strange because I am sure I am not using link.exe at all. I bootstrapped `vcpkg` under MSYS2's Mingw64 shell, using `g++` to compile `vcpkg` itself etc. And I am sure Visual C++ toolchain is not in the PATH for Mingw64 shell. 

So, if second linker member is only generated if VS link.exe is used, how then I was able to build other ports without complain? They didn't use link.exe either but they passed the post-build checks just fine.

**Comment by Neumann-A** [comment]
because there is a policy setting in the triplet. But this policy setting only works for import libraries/dll not for static libraries.

**Comment by sfhacker** [comment]
@thumpro Do not mix environments and/or compilers and/or linkers (e.g. link, cl, g++, etc).
You are looking for trouble!

**Comment by thumpro** [comment]
@sfhacker I can assure you, I am not mixing toolchains in this issue. The whole idea of the issue I first linked, about bootstrapping vcpkg with Mingw64, is to be able to bootstrap and use `vcpkg` without even installing Visual Studio. 

@Neumann-A I tried downloading the `szip` source and build it manually through their configure and Makefile. configure recognized the system as mingw64 and ended up building a static library. There was a warning error msg, somewhere around the lines `Undefined symbols are not supported by Mingw dlls, building static library instead`. So, it seems there is an issue with natively building this library as shared library under Mingw. And CMake somehow manages it but it does not pass the dumpbin checks in the end.

To that end, I added you policy, first directly to `x64-mingw-dynamic` triplet cmake file, `set(VCPKG_POLICY_SKIP_DUMPBIN_CHECKS enabled)`. This didn't help, second linker member check is still performed and it failed. 

Then I tried using the policy in szip's `portfile.cmake`, as follows:
    if(${TARGET_TRIPLET}  MATCHES "x*-mingw")
        set(VCPKG_POLICY_SKIP_DUMPBIN_CHECKS enabled)
    endif()

This does not help either. It looks like it should definitely bypass dumpbin checks but I am still getting the `Cannot find proper second linker member`. And surely, my software is recent enough to include the commit where you introduced that policy.
https://github.com/microsoft/vcpkg/pull/9901/files/9298dec358d145b0feb0d7336f755c639fbacde8

I guess the other policy you were talking about, the one in the triplet file was `set(VCPKG_POLICY_DLLS_WITHOUT_LIBS enabled)`. For `szip` this triplet creates the .lib file, .a file and .dll. But as we said, .lib's dumpbin check fails. All the other ports that I built successfully for Mingw64 created only a .dll file and a .dll.a file. I have no idea why this one creates a .lib file as well which makes the whole check fail.

**Comment by Neumann-A** [comment]
try 
```
set(VCPKG_POLICY_SKIP_ARCHITECTURE_CHECK enabled)
set(VCPKG_POLICY_SKIP_DUMPBIN_CHECKS enabled)
```
it should work since the same problem arises if compiling with clang-cl and using ldd

**Comment by sfhacker** [comment]
@thumpro It seems we are unable to make any progress here! We are stuck, going round in circles!
You lost me here!

**Comment by thumpro** [comment]
@sfhacker Don't worry, we are actually making quite the progress 👍

```
set(VCPKG_POLICY_SKIP_ARCHITECTURE_CHECK enabled)
set(VCPKG_POLICY_SKIP_DUMPBIN_CHECKS enabled)
```

with this, it actually worked. Could you explain the machinery a little bit @Neumann-A and what is actually going on here? Since I bypassed the check essentially, I am not 100% sure in my mind that the build was correct and I won't be lacking any required smybols etc. inside the built binaries. 

The CMake targets that were suggested to use at the end was:
```
find_package(szip CONFIG REQUIRED)
    target_link_libraries(main PRIVATE szip-shared szip-static)
```

**Comment by Neumann-A** [comment]
Mingw building both targets? you should only link against one of them.

the checks within vcpkg are only performed on windows and under the assumption that MSVC is used. As such disabling them for other things is fine. The linker will warn you otherwise if the architecture is wrong.

**Comment by thumpro** [comment]
Then how other people using this `x64-mingw-dynamic` triplet are able to build this port without adding that policy to their `portfile.cmake`? For example, @sfhacker shows above a screenshot showing a successful build of `szip`. 

If I use
> .\bootstrap-vcpkg.bat

to bootstrap my `vcpkg` and then use `vcpkg.exe install some_port:x64-mingw-dynamic`, will it still use Visual C++ toolchain to build stuff? Because then you would not be able to link those with mingw64 distributed `g++` since they are built with `MSVC`. In a sense, I am uncertain how people are using this `x64-mingw-dynamic` triplet on a Windows machine, bootstrapped using the .bat bootstrap script.

**Comment by longnguyen2004** [comment]
./bootstrap-vcpkg.bat only uses MSVC to build vcpkg itself. Libraries will be built with mingw if you use the mingw triplets (see https://github.com/microsoft/vcpkg/blob/master/scripts/toolchains/mingw.cmake)

**Comment by thumpro** [comment]
@longnguyen2004 Just one more thing that I really wanted to know, thanks. I guess since this is a community triplet, there is not much documentation.

**Comment by longnguyen2004** [comment]
@thumpro It all depends on the community to provide support for it. I've worked on a lot of other ports (openssl, libffi, freetype, just to name a few) to get it to build on mingw. Anyway I'll take a look at szip.

**Comment by longnguyen2004** [comment]
@thumpro Please test #14460

**Comment by thumpro** [comment]
Did a clean checkout of this branch and verified that szip builds. Thanks a lot! 

Then I moved on to building the package that I actually needed szip for, hdf5, to see if it actually correctly links with other packages. There I was not able to move further because `zlib` dependency started failing. I am not sure if your changes can cause zlib port to fail and I can not directly say that is the reason but zlib was building fine for me previously. And this is just a clean checkout of your branch.

**Comment by longnguyen2004** [comment]
Can you upload the build log? It's building fine for me
<details>
<summary>
Log
</summary>

```
./vcpkg --triplet x64-mingw-dynamic install zlib
Your feedback is important to improve Vcpkg! Please take 3 minutes to complete our survey by running: vcpkg contact --survey
Computing installation plan...
The following packages will be built and installed:
    zlib[core]:x64-mingw-dynamic
Detecting compiler hash for triplet x64-mingw-dynamic...
Starting package 1/1: zlib:x64-mingw-dynamic
Building package zlib[core]:x64-mingw-dynamic...
Could not locate cached archive: C:\Users\[redacted]\AppData\Local\vcpkg\archives\00\00221dc6ee21e0aee1145088a50007e348759eb7.zip
-- Using community triplet x64-mingw-dynamic. This triplet configuration is not guaranteed to succeed.
-- [COMMUNITY] Loading triplet configuration from: C:\vcpkg\triplets\community\x64-mingw-dynamic.cmake
-- Using cached C:/vcpkg/downloads/zlib1211.tar.gz
-- Cleaning sources at C:/vcpkg/buildtrees/zlib/src/1.2.11-63309e48e5.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/vcpkg/downloads/zlib1211.tar.gz
-- Applying patch cmake_dont_build_more_than_needed.patch
-- Applying patch 0001-Prevent-invalid-inclusions-when-HAVE_-is-set-to-0.patch
-- Applying patch add_debug_postfix_on_mingw.patch
-- Using source at C:/vcpkg/buildtrees/zlib/src/1.2.11-63309e48e5.clean
-- Configuring x64-mingw-dynamic
-- Building x64-mingw-dynamic-dbg
-- Building x64-mingw-dynamic-rel
-- Using msys root at C:/vcpkg/downloads/tools/msys2/c809757c94447846
-- Fixing pkgconfig file: C:/vcpkg/packages/zlib_x64-mingw-dynamic/lib/pkgconfig/zlib.pc
-- Fixing pkgconfig file: C:/vcpkg/packages/zlib_x64-mingw-dynamic/debug/lib/pkgconfig/zlib.pc
-- Installing: C:/vcpkg/packages/zlib_x64-mingw-dynamic/share/zlib/copyright
-- Performing post-build validation
-- Performing post-build validation done
Stored binary cache: C:\Users\[redacted]\AppData\Local\vcpkg\archives\00\00221dc6ee21e0aee1145088a50007e348759eb7.zip
Building package zlib[core]:x64-mingw-dynamic... done
Installing package zlib[core]:x64-mingw-dynamic...
Installing package zlib[core]:x64-mingw-dynamic... done
Elapsed time for package zlib:x64-mingw-dynamic: 2.258 min

Total elapsed time: 3.618 min

The package zlib is compatible with built-in CMake targets:

    find_package(ZLIB REQUIRED)
    target_link_libraries(main PRIVATE ZLIB::ZLIB)

```
</details>

**Comment by sfhacker** [comment]
I have seen several issues regarding zlib!

**Comment by thumpro** [comment]
About zlib, it is definitely not related to this issue, but something else. During ninja build, it looks for a program called `x86_64-w64-mingw32-windres.exe` due to some CMake config or something I guess. The problem is, this binary was named just `windres.exe` on my MinGW installation which I installed through MSYS2. I created a copy of it with the correct name and was able to build zlib fine. Still, this might use some streamlining.

**Comment by longnguyen2004** [comment]
I wasn't aware that prefixed `windres` isn't always available, I'll add a fallback into the toolchain file.

**Comment by thumpro** [comment]
Cool, I am looking forward to merging of this PR.

**Comment by thumpro** [comment]
Maybe we should close this ticket once the other PR is merged but since we caught some traction on this community triplet, I felt like adding this to here.

`ffmpeg` with x64-mingw-dynamic community triplet was failing for me (which is apart from hdf5, the other big dependency for opencv4 if you want to do video capture etc.)

I had to change the portfile as:

`$ git diff ports/ffmpeg/portfile.cmake
diff --git a/ports/ffmpeg/portfile.cmake b/ports/ffmpeg/portfile.cmake
index cbe8df318..f09e57df4 100644
--- a/ports/ffmpeg/portfile.cmake
+++ b/ports/ffmpeg/portfile.cmake
@@ -25,7 +25,7 @@ endif()
 vcpkg_find_acquire_program(YASM)
 get_filename_component(YASM_EXE_PATH ${YASM} DIRECTORY)

-if(VCPKG_TARGET_IS_WINDOWS)
+if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)
     #We're assuming that if we're building for Windows we're using MSVC
     set(INCLUDE_VAR "INCLUDE")
     set(LIB_PATH_VAR "LIB")
@@ -37,6 +37,10 @@ endif()
 set(ENV{PATH} "$ENV{PATH}${VCPKG_HOST_PATH_SEPARATOR}${YASM_EXE_PATH}")
 set(OPTIONS "--enable-asm --enable-yasm --disable-doc --enable-debug --enable-runtime-cpudetect")

+if(VCPKG_TARGET_IS_MINGW)
+       set(VCPKG_TARGET_IS_WINDOWS false)
+endif()
+
 if(VCPKG_TARGET_IS_WINDOWS)
     if(VCPKG_TARGET_ARCHITECTURE STREQUAL "arm" OR VCPKG_TARGET_ARCHITECTURE STREQUAL "arm64")
         vcpkg_acquire_msys(MSYS_ROOT
@@ -61,8 +65,15 @@ if(VCPKG_TARGET_IS_WINDOWS)

     set(SHELL ${MSYS_ROOT}/usr/bin/bash.exe)
     set(OPTIONS "--toolchain=msvc ${OPTIONS}")
-else()
-    set(SHELL /bin/sh)
+elseif(VCPKG_TARGET_IS_MINGW)
+       set(SHELL sh)
+       if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+               set(OPTIONS "${OPTIONS} --target-os=mingw64")
+       elseif(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86")
+               set(OPTIONS "${OPTIONS} --target-os=mingw32")
+       endif()
+else()
+       set(SHELL /bin/sh)
 endif()

 set(ENV{${INCLUDE_VAR}} "${CURRENT_INSTALLED_DIR}/include${VCPKG_HOST_PATH_SEPARATOR}$ENV{${INCLUDE_VAR}}")`

to finally be able to build. Also, that port is not printing usage information once the build is done. That might be due to my changes.

**Comment by longnguyen2004** [comment]
Hmm weird, I remember it working a while ago, I'll take a look at some point
Also, do not unset VCPKG_TARGET_IS_WINDOWS, instead use if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)

**Comment by dg0yt** [comment]
> So, if second linker member is only generated if VS link.exe is used, how then I was able to build other ports without complain? They didn't use link.exe either but they passed the post-build checks just fine.

For the record, the difference is whether import library generated by the mingw linker is named "....dll.a" (the common case with mingw) or "....lib" (the szip case). The failing check in vcpkg-tool is applied only to ".lib" files. 

#14460 changes the policy in the mingw triplet just for the szip. I wonder how this is meant to scale. I run into this error with hdf5 today. I can submit another if/endif, or the policy is generally toggled in the mingw triplets, due to "second linker member are only generated if VS link.exe is used".

**Comment by longnguyen2004** [comment]
After more investigation, I realized the `.lib` file isn't a proper VS import library, but a mingw import library with the extension changed to `.lib`. We can either ask upstream to change the behavior, or patch it out by ourselves, which is what MSYS2 has done (`libszip`: [here](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-szip/szip-default-import-suffix.patch) and [here](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-szip/szip-proper-library-names-mingw.patch), `hdf5`: [here](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-hdf5/hdf5-default-import-suffix.patch) and [here](https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-hdf5/hdf5-proper-library-names-mingw.patch))

**Comment by longnguyen2004** [comment]
I'll pull the patch from MSYS2 and make a pull request when I have time.
